### PR TITLE
[MM-1915] Add Deactivate Account  - reborn :)

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
Add an option to user deactivated his account

#### Ticket Link
Jira: https://mattermost.atlassian.net/browse/MM-1915

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes - https://github.com/mattermost/mattermost-server/pull/8699
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates

### CODE REVIEW

**Comment by esethna** [comment]
Thanks Carlos. @amyblais will be handling the first PM review when ready

**Comment by cpanato** [comment]
Server PR was merge.

**Comment by amyblais** [comment]
@AndersonWebStudio Can you help turn on "EnableUserDeactivation" for the test server?

**Comment by amyblais** [comment]
In the [corresponding server PR](https://github.com/mattermost/mattermost-server/pull/8699), there is the following config setting added: https://github.com/mattermost/mattermost-server/pull/8699/files#diff-1e9c3d615e9ebaaaa3669b4c2fd87d00

Perhaps it wasn't included in this PR, @cpanato can you take a look?

**Comment by AndersonWebStudio** [comment]
@amyblais thanks I think I found it, setting is updated.

**Comment by amyblais** [comment]
I still don't see the setting and I realize t's probably because of the failing checks, @cpanato can you help resolve the failing checks on this PR?

**Comment by cpanato** [comment]
@amyblais rebased and set the flag `EnableUserDeactivation` to true for testing.

**Comment by amyblais** [comment]
@esethna When testing this, can you focus on #2 below:

Along with the Jira ticket description, these are also part of this PR:

a) config.json setting.
b) email notification (received when a person deactivates their account so that they know the deactivation succeeded).

**Comment by esethna** [comment]
@cpanato sorry for the delay in reviewing this PR, can you please help rebase, set up the test server then turn on EnableUserDeactivation?

**Comment by cpanato** [comment]
@esethna rebased and the test server I set the `EnableUserDeactivation` to `true`

**Comment by esethna** [comment]
@cpanato apologies this got missed for review. Can you please retrigger the spinmint and set the EnableUserDeactivation to true?

**Comment by cpanato** [comment]
@esethna Done! 😄

**Comment by esethna** [comment]
@wiersgallak test server is up if you would like to do a first pass review on this PR?

**Comment by wiersgallak** [comment]
Looks good-works as outlined on the ticket @cpanato.  Tested opening the module, canceling the module, deactivating account, trying to login when deactivated, viewed as system admin, tried to DM as sys admin, and reactivating account and verified account was again active.  

Only thing that stuck out to me was we may want to consider some consistency in the description "Open to deactivate your account".  On the General tab, we use "Click 'Edit' to add your nickname" and "Click 'Edit' to upload your picture" , on the Display tab we have "Open to manage your theme".

**Comment by wiersgallak** [comment]
Also verified the notification worked as expected once I deactivated the account.

**Comment by cpanato** [comment]
@wiersgallak what do you suggest?

**Comment by wiersgallak** [comment]
@cpanato I would suggest updating all to be consistent with "Click 'Edit' to" [action].  You can open the modal by just clicking on the UI, but we do have the edit menu, so we should use the visual cue to direct our users.

**Comment by cpanato** [comment]
@wiersgallak did the change and there is a new test server to check 😄

**Comment by wiersgallak** [comment]
Looks good @cpanato - Thanks!

**Comment by jasonblais** [comment]
@wiersgallak Is this good from PM side? If so, we can 
- remove the test server and PM review labels
- add dev review label with two reviewers

File: components/user_settings/advanced/user_settings_advanced.jsx [hunk 1]
```diff
@@ -509,6 +551,93 @@ export default class AdvancedSettingsDisplay extends React.Component {
             }
         }
 
+        let deactivateAccountSection = '';
+        let makeConfirmationModal = '';
+        const currentUser = UserStore.getCurrentUser();
+        const config = getConfig(store.getState());
```
Comment (saturninoabril) [inline components/user_settings/advanced/user_settings_advanced.jsx]: We have `config` on [index.js](https://github.com/cpanato/mattermost-webapp/blob/eb7d7287136b62793843bc391feea7d4ad52034c/components/user_settings/advanced/index.js#L10).  I think we can get the `EnableUserDeactivation` from there and just pass as props.

File: components/confirm_modal.jsx [hunk 2]
```diff
@@ -175,6 +175,7 @@ export default class ConfirmModal extends React.Component {
                 <Modal.Footer>
                     {cancelButton}
                     <button
+                        autoFocus='true'
```
Comment (lieut-data) [inline components/confirm_modal.jsx]: This is a site-wide change, right? Have we confirmed we want this behaviour? What's this fixing, precisely?

File: components/user_settings/advanced/user_settings_advanced.jsx [hunk 3]
```diff
@@ -134,6 +142,40 @@ export default class AdvancedSettingsDisplay extends React.Component {
...
+        );
+    }
+
+    handleShowDeactivateAccountModal = () => {
+        this.setState({
+            showDeactivateAccountModal: true,
+        });
+    }
+
+    handleDeactivateAccountCancelModal = () => {
```
Comment (lieut-data) [inline components/user_settings/advanced/user_settings_advanced.jsx]: Should this be called something like `handleHideDeactivateAccountModal` to mirror above?

File: components/confirm_modal.jsx [hunk 4]
```diff
@@ -85,19 +85,19 @@ export default class ConfirmModal extends React.Component {
 
     componentDidMount() {
         if (this.props.show) {
-            document.addEventListener('keypress', this.handleKeypress);
+            document.addEventListener('keydown', this.handleKeypress);
```
Comment (lieut-data) [inline components/confirm_modal.jsx]: This change seems orthogonal to the new feature -- what's the context?

Comment (cpanato) [reply components/confirm_modal.jsx]: as far I remember, @saturninoabril and I checked that and to use the enter key across modal we made this change because the enter key was not captured by the modal and was captured by another component. 
let me know if you want to change that.

File: components/confirm_modal.jsx (hunk 2 above)
Comment (cpanato) [reply components/confirm_modal.jsx]: was requested to get the autofocus

File: components/user_settings/advanced/user_settings_advanced.jsx (hunk 3 above)
Comment (cpanato) [reply components/user_settings/advanced/user_settings_advanced.jsx]: done, thanks!

File: components/confirm_modal.jsx (hunk 4 above)
Comment (saturninoabril) [reply components/confirm_modal.jsx]: That's right.  It's weird that pressing `enter` key works on `keydown` and not on `keypress` in this component.

File: components/confirm_modal.jsx (hunk 2 above)
Comment (lieut-data) [reply components/confirm_modal.jsx]: I think we should tease this and the `keydown` behaviour out of this PR and submit it separately. My concern is that, in the event of unexpected side effect, this PR doesn't "stand out" as a cause unless one looks deep into the commits. Even then, it would be annoying to revert the whole thing instead of just the orthogonal changes here.

Comment (cpanato) [reply components/confirm_modal.jsx]: ok I will remove this and submit in another PR

Comment (cpanato) [reply components/confirm_modal.jsx]: done

Comment (cpanato) [reply components/confirm_modal.jsx]: will open another PR only with this change

Comment (lieut-data) [reply components/confirm_modal.jsx]: 0/5, but wondering if we should push `autoFocus` too. Otherwise I've approved the PR :)

**Comment by jasonblais** [comment]
@cpanato Katie tested it instead so I think this is good to merge - though propose we wait until AFTER 5.2 release branch is cut on Friday since the feature merge deadline was on Monday.

**Comment by cpanato** [comment]
ok sounds good @jasonblais
